src/afrolabs/components/kafka.clj

```diff
@@ -317,7 +333,70 @@
        (apply ~n (rest strategy-specs#)))))
 
 
-;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defstrategy RenameTopicsForProducer
+  [& {:keys [rename-fn]}]
+  (when-not rename-fn
+    (throw (ex-info (str "strategy RenameTopicsForProducer requires the `rename-fn` to be specified. This function "
+                         "accepts a topic name for every message, and must return a modified version of the topic, "
+                         "OR nil, in which case the original topic value will be kept.")
+                    {})))
+  (reify
+    IProducerPreProduceMiddleware
+    (pre-produce-hook [_ msgs]
+      (map (fn [{:as   msg
+                 :keys [topic]}]
+             (assoc msg :topic (or (rename-fn topic)
+                                   topic)))
+           msgs))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn- resolve-fn
+  "Resolves a configuration value that semantically points to an fn, to the actual fn.
+  `the-fn` might be a symbol, in which case the namespace will be recquired and the symbol resolved
+  or an actual `IFn`."
```

Comment:
This should mention that it returns `identity` if the function is not found. However, would it not be better to throw in that case? Both of the places I see this used already throw on `nil` rather than using the identity.

On further reflection returning `identity` breaks the implied contract of the name in a way I would find very surprising if I came across it in the wild.  Unless you are allowing this function to be set dynamically at runtime I think throwing here will fail faster and force you to fix broken references.

Reply:
I've changed it around a little bit:
- `UpdateConsumedRecordKey` will throw if it was passed a nil-valued parameter
- `resolve-fn` will throw if it cannot resolve the reference to a non-nil value.